akvo/rest/viewsets.py

```diff
@@ -11,15 +11,40 @@
 
 from .models import TastyTokenAuthentication
 
-from akvo.rsr.models import Project
-
 
 class BaseRSRViewSet(viewsets.ModelViewSet):
     """
     Base class used for the view sets for RSR models. Provides unified auth and perms settings.
-    Only public projects will be shown by filtering the queryset.
     """
     authentication_classes = (SessionAuthentication, TastyTokenAuthentication, )
     permission_classes = (DjangoObjectPermissions, )
     filter_backends = (filters.DjangoFilterBackend, filters.OrderingFilter, )
     ordering_fields = '__all__'
+
+
+class PublicProjectViewSet(BaseRSRViewSet):
+    """
+    Only public projects or objects related to public projects will be shown.
+    """
+    def get_queryset(self, related_to='project__'):
+        queryset = super(PublicProjectViewSet, self).get_queryset()
+        user = self.request.user
+
+        if not (user.is_superuser or user.is_admin):
+            project_filter = related_to + 'is_public'
+
+            public_objects = queryset.filter(**dict.fromkeys([project_filter], True)).distinct()
+            private_objects = queryset.filter(**dict.fromkeys([project_filter], False)).distinct()
+
+            for obj in private_objects:
+                # Loop through all 'private' objects to see if the user has permission to change it
+                # or not. Fetch permission first. E.g. a Sector object becomes 'rsr.change_sector'.
+                permission = type(obj)._meta.db_table.replace('_', '.change_')
```

Comment:
Add a limit of one replacement to safeguard against model names with "_" in them:
```python
permission = type(obj)._meta.db_table.replace('_', '.change_', 1)
```
Don't think we have any in RSR, except for the hidden many-to-many tables, but just to be safe...